src/lucky_router/fragment.cr

```diff
@@ -59,12 +59,24 @@ class LuckyRouter::Fragment(T)
   # the final step to finding the payload is to search for a matching request method
   getter method_to_payload = Hash(String, T).new
 
+  def find(parts : Array(String), method : String) : Match(T) | NoMatch
+    MatchFinder(T).new(self, parts: parts, method: method).run
+  end
+
   def process_parts(parts : Array(String), method : String, payload : T)
-    PartProcessor(T).new(self, parts: parts, method: method, payload: payload).run
-    self
+    leaf_fragment = parts.reduce(self) { |fragment, part| fragment.add_part(part) }
+    leaf_fragment.method_to_payload[method] = payload
   end
 
-  def find(parts : Array(String), method : String) : Match(T) | NoMatch
-    MatchFinder(T).new(self, parts: parts, method: method).run
+  def add_part(part : String) : Fragment(T)
+    if part.starts_with?(":")
```

Comment:
Does this also handle [optional parts](https://github.com/luckyframework/lucky_router/blob/master/spec/integration_spec.cr#L49)? I see the specs are passing, but it would seem if a part started with `?:` then this wouldn't be picked up here... Or am I missing where that logic happens?

Reply:
Good question, after looking into it, optional parts are handled in the matcher https://github.com/luckyframework/lucky_router/blob/73148013bc1e6848f9c1c1342c4e5961a43f9faf/src/lucky_router/matcher.cr#L28-L32

Reply:
That's interesting. I haven't dug in to the router as much as you have, do you think that's fine?

On a side note, I just noticed 2 tiny optimizations. `part.starts_with?(':')` or `part.starts_with?('?')`. And `part.delete('?')`. It would need to be tested, but I think using chars and the use of `delete` over gsub are "better".

Reply:
In the next PR or so, I'm going to be wrapping the path and path part strings in actual classes. I think that will allow us to move the optional parts down into the fragment logic and will give an excuse to make these optimizations.

Reply:
Awesome, that sounds good then!

Reply:
I like that idea @matthewmcgarvey and @jwoertink. Just want to note that optimization in the part processor won't buy us much since it only process once when the app boots and takes less than a millisecond even for lots of routes so doesn't benefit *as* much from performance optimization. But since `delete` makes more sense anyway we might as well do that in the next PR!